File: kodex-codebuddy/src/openai-types.ts
/**
 * Minimal OpenAI Chat Completions request/response types we need to handle.
 * Only the subset that this proxy supports; unknown fields are ignored.
 */

export interface OAIMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string | OAIContentPart[] | null;
  /** Present when role === 'tool' (result of a previous tool call). */
  tool_call_id?: string;
  /** Present when role === 'assistant' and the model emitted tool calls. */
  tool_calls?: OAIToolCall[];
  name?: string;
}

export interface OAIContentPart {
  type: 'text' | 'image_url';
  text?: string;
  image_url?: { url: string; detail?: string };
}

export interface OAIToolCall {
  id: string;
  type: 'function';
  function: { name: string; arguments: string };
}

export interface OAITool {
  type: 'function';
  function: {
    name: string;
    description?: string;
    parameters?: Record<string, unknown>;
  };
}

export interface OAIChatRequest {
  model?: string;
  messages: OAIMessage[];
  stream?: boolean;
  temperature?: number;
  max_tokens?: number | null;
  max_completion_tokens?: number | null;
  tools?: OAITool[];
  tool_choice?: 'none' | 'auto' | 'required' | { type: 'function'; function: { name: string } };
  top_p?: number;
  stop?: string | string[];
  user?: string;
  /** Other fields passed through; ignored by the proxy. */
  [key: string]: unknown;
}

// --- Response shapes ---

export interface OAIChoiceMessage {
  role: 'assistant';
  content: string | null;
  tool_calls?: OAIToolCall[];
  refusal?: string | null;
}

export interface OAIChoice {
  index: number;
  message: OAIChoiceMessage;
  finish_reason: 'stop' | 'length' | 'tool_calls' | 'content_filter' | null;
  logprobs?: null;
}

export interface OAIChatResponse {
  id: string;
  object: 'chat.completion';
  created: number;
  model: string;
  choices: OAIChoice[];
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
  system_fingerprint?: string;
}

// --- Streaming chunk shapes ---

export interface OAIChunkDelta {
  role?: 'assistant';
  content?: string | null;
  tool_calls?: Array<{
    index: number;
    id?: string;
    type?: 'function';
    function: { name?: string; arguments?: string };
  }>;
  refusal?: string | null;
}

export interface OAIChunkChoice {
  index: number;
  delta: OAIChunkDelta;
  finish_reason: 'stop' | 'length' | 'tool_calls' | 'content_filter' | null;
}

export interface OAIChatChunk {
  id: string;
  object: 'chat.completion.chunk';
  created: number;
  model: string;
  choices: OAIChunkChoice[];
}

// --- Models list ---

export interface OAIModel {
  id: string;
  object: 'model';
  created: number;
  owned_by: string;
}

export interface OAIModelsResponse {
  object: 'list';
  data: OAIModel[];
}

File: crates/codebuddy-proxy/src/openai_types.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiMessage {
    pub role: String,
    #[serde(default)]
    pub content: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiTool {
    #[serde(rename = "type")]
    pub type_field: String,
    pub function: OaiToolFunction,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiToolFunction {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parameters: Option<Value>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChatRequest {
    #[serde(default)]
    pub model: Option<String>,
    pub messages: Vec<OaiMessage>,
    #[serde(default)]
    pub stream: Option<bool>,
    #[serde(default)]
    pub tools: Option<Vec<OaiTool>>,
    #[serde(default)]
    pub max_tokens: Option<Value>,
    #[serde(default)]
    pub temperature: Option<f64>,
    #[serde(flatten)]
    pub extra: Value,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChatResponse {
    pub id: String,
    pub object: String,
    pub created: u64,
    pub model: String,
    pub choices: Vec<OaiChoice>,
    pub usage: OaiUsage,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChoice {
    pub index: u32,
    pub message: OaiChoiceMessage,
    pub finish_reason: String,
    pub logprobs: Option<Value>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChoiceMessage {
    pub role: String,
    pub content: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<Value>>,
}
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct OaiUsage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
    /// OpenAI-style nested cache hit count. `codex_api_proxy` reads this via
    /// `prompt_tokens_details.cached_tokens`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prompt_tokens_details: Option<OaiPromptTokensDetails>,
    /// Anthropic-style cache-read count. Kept as a top-level alias so
    /// `usage_cached_input_tokens` can pick it up without nested details.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cache_read_input_tokens: Option<u64>,
    /// Anthropic-style cache-write / creation count.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cache_creation_input_tokens: Option<u64>,
}

impl OaiUsage {
    pub fn zero() -> Self {
        Self {
            prompt_tokens: 0,
            completion_tokens: 0,
            total_tokens: 0,
            prompt_tokens_details: None,
            cache_read_input_tokens: None,
            cache_creation_input_tokens: None,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct OaiPromptTokensDetails {
    pub cached_tokens: u64,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChatChunk {
    pub id: String,
    pub object: String,
    pub created: u64,
    pub model: String,
    pub choices: Vec<OaiChunkChoice>,
    /// OpenAI streams a terminal `choices: []` chunk carrying `usage` when
    /// `stream_options.include_usage` is set. We emit it unconditionally
    /// (internal reverse proxy for codex) so clients can account per-turn
    /// tokens even without opting in.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub usage: Option<OaiUsage>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiChunkChoice {
    pub index: u32,
    pub delta: Value,
    pub finish_reason: Option<String>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OaiModelsResponse {
    pub object: String,
    pub data: Vec<Value>,
}
